verification/spec/relayer/environment.tla

```diff
@@ -0,0 +1,275 @@
+    \union
+    [type : {"ClientUpdate"}, clientID : ClientIDs, height : Heights]    
+
+(***************
+Client ID operators
+***************)
+
+\* given a clientID, translate back to srcChainID and dstChainID
+GetChainIDs(clientID) ==
+    LET srcChainID == (clientID - 1) \div NrChains + 1 IN
+    LET dstChainID == ((clientID - 1) % NrChains) + 1 IN
+    [src |-> srcChainID, dst |-> dstChainID]
+    
+\* get the clientID of the client on the chain onChainID, for the chain forChainID
+GetClientID(onChainID, forChainID) ==
+    (onChainID - 1) * NrChains + forChainID
+    
+\* check if the clientID is a valid client ID w.r.t. chainID    
+ValidClientID(chainID, clientID) ==
+    /\ GetChainIDs(clientID).src = chainID
+    /\ GetChainIDs(clientID).dst \in ChainIDs 
+    
+(***************
+Chain update operators
+***************)    
+
+LightClientUpdate(chainID, chain, datagrams) == 
+    \* update chain with "CreateClient" datagrams
+    LET createClients == {dgr \in datagrams : dgr.type = "CreateClient"} IN
+    LET createClientIDs == {clientID \in ClientIDs : 
+            /\ \E dgr \in createClients : dgr.clientID = clientID
+            /\ ValidClientID(chainID, clientID)} IN
+    LET createClientHeights(clientID) == 
+            {h \in Heights : 
+                \E dgr \in createClients : 
+                    dgr.clientID = clientID /\ dgr.height = h} IN  
+    
+    \* new chain record with clients created
+    LET clientsCreated == [
+            height |-> chain.height,
+            clientHeights |-> 
+                [cID \in ChainIDs |-> IF GetClientID(chainID, cID) \in createClientIDs
+                                      THEN Max(createClientHeights(GetClientID(chainID, cID)))
+                                      ELSE chain.clientHeights[cID] 
+                ]
+         ] IN
+     
+    \* update chain with "ClientUpdate" datagrams
+    LET updateClients == {dgr \in datagrams : dgr.type = "ClientUpdate"} IN
+    LET updateClientIDs == {clientID \in ClientIDs : 
+            /\ \E dgr \in updateClients : dgr.clientID = clientID
+            /\ ValidClientID(chainID, clientID)} IN
+    LET updateClientHeights(clientID) == 
+            {h \in Heights : \E dgr \in updateClients : 
+                dgr.clientID = clientID /\ dgr.height = h} IN  
+    
+    \* new chain record with clients updated
+    LET clientsUpdated == [
+            height |-> clientsCreated.height,
+            clientHeights |-> 
+                [cID \in ChainIDs |-> IF GetClientID(chainID, cID) \in updateClientIDs
+                                      THEN Max(updateClientHeights(GetClientID(chainID, cID)))
+                                      ELSE clientsCreated.clientHeights[cID] 
+                ]
+         ] IN
+    
+    clientsUpdated
+ 
+\* Update chainID with the received datagrams
+\* Currently, only supporting ICS2: Client updates
+\* TODO: Support the remaining datagrams
+UpdateChain(chainID, datagrams) == 
```

Comment:
I realised that TLA+ specs should be read bottom up as high level functions come after helper functions are defined :)